dask/array/routines.py

```diff
@@ -332,6 +333,45 @@ def vdot(a, b):
     return dot(a.conj().ravel(), b.ravel())
 
 
+def _chunk_sum(a, axis=None, dtype=None, keepdims=None):
+    # Caution: this is not your conventional array-sum: due
+    # to the special nature of the preceding blockwise con-
+    # traction,  each chunk is expected to have exactly the
+    # same shape,  with a size of 1 for the dimension given
+    # by `axis` (the reduction axis).  This makes mere ele-
+    # ment-wise addition of the arrays possible.   Besides,
+    # the output can be merely squeezed to lose the `axis`-
+    # dimension when keepdims = False
+    if type(a) is list:
+        xp = np
+
+        if is_cupy_type(a[0]):
+            import cupy
+
+            xp = cupy
+
+        out = reduce(partial(xp.add, dtype=dtype), a)
```

Comment:
```suggestion
        out = reduce(partial(np.add, dtype=dtype), a)
```

Reply:
Thank you @pentschev 

One final question: what about the following code-block already in the dask source-code (but also directly related to this PR)?
```python
def _matmul(a, b):
    xp = np

    if is_cupy_type(a):
        import cupy
        xp = cupy

    chunk = xp.matmul(a, b)
```
Can it also be similarly replaced as you did?
```python
    chunk = np.matmul(a, b)
```

Reply:
@ParticularMiner I **believe** it could be removed **BUT** only assuming certain cupy version (>=9.0), see https://github.com/cupy/cupy/issues/4549. I assume that assumption is not "safe", 9.0 was relatively recent. I think if I was you, would add a comment there to the cupy issue + this discussion + explanation that it probably can be eventually removed(?) Unless of course we dask is comfortable with failing there and forcing cupy upgrade, which is probably not worth the potential bad user experience.

Reply:
@ravwojdyla Very interesting link!  I wasn't aware of the `__array_ufunc__` protocol.  Yes, I completely agree with you that we should beware of possible backward compatibility issues here.  So I'll leave the code as it is and include a comment, as you suggest.